### PR TITLE
Feat: suppot TFGrid-KYC 

### PR DESCRIPTION
**Pull request by 0oM4R**
### Description

Add KYC client to grid client, dashboard, and types packages.

changes: 

- grid client:
   - add kyc client
   - add kyc client to grid client
   - check if kyc verified on create name and handle methods in twinddeployment 
   
- types:  add related error classes 

- dashboard:
   - add kyc verifier that uses iframe to connect to idenfy
   - add kyc store, init on profile, clear if no profile
   - prevent deployment if kyc is not verified, in view layout
 
### tested scenarios:
 - try to deploy without kyc verification all deployments should be blocked
 - check twin page to see the kyc status
 - use kyc client and check its methods
 - check error handling in dashboard, toast etc, you can manipulate the headers in kyc client
 - check if you can deploy without kyc in gridclient
 - try to enter the verification process then cancel 
 - try to have rejected kyc verification
 - try to verify the kyc with valid data using the dashboard
 - try to scan the qr code  and continue verification process on mobile 
- make sure that the kyc status got cleaned after logout/refresh

related issues: 

- #3519 

info issue: https://github.com/threefoldtech/info_grid/issues/666

screenshots: 
Screenshot from 2024-10-23 12-21-41
Screenshot from 2024-10-23 12-21-33
image
Screenshot from 2024-10-23 12-27-40
Screenshot from 2024-10-23 12-27-19
image
image

### CODE REVIEW

File: packages/grid_client/src/clients/kyc/client.ts [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+  /**
+   * Creates an instance of KYC.
+   * @param apiDomain - The API domain for the TFGrid KYC service.
+   * @param address - The TFChain user address.
+   * @param keypairType - The type of keypair to use (default is sr25519).
+   * @param mnemonic - The mnemonic for generating the keypair.
+   */
+  constructor(
+    public apiDomain: string,
+    public address: string,
```
Comment (AhmedHanafy725) [inline packages/grid_client/src/clients/kyc/client.ts]: why do we need an address here? you can get it from the mnemonic

File: packages/grid_client/src/clients/kyc/client.ts [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+  private async getKey(): Promise<KeyringPair> {
+    await waitReady();
+    return this.keyr.addFromUri(this.mnemonic);
+  }
+  /**
+   * Converts a string message to its hexadecimal representation.
+   * @param message - The message to convert.
+   * @returns The hexadecimal representation of the message.
+   */
+  private stringToHex(message: string): string {
```
Comment (AhmedHanafy725) [inline packages/grid_client/src/clients/kyc/client.ts]: this should be in utils

File: packages/grid_client/src/clients/kyc/client.ts [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    return Buffer.from(message).toString("hex");
+  }
+  /**
+   * Converts a hexadecimal string to a byte array.
+   * @param hex - The hexadecimal string to convert.
+   * @returns {Uint8Array} The byte array representation of the hexadecimal string.
+   */
+  private bytesFromHex(hex: string): Uint8Array {
```
Comment (AhmedHanafy725) [inline packages/grid_client/src/clients/kyc/client.ts]: this one too should be in utils

File: packages/grid_client/src/clients/kyc/client.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+  /**
+   * Fetches the verification data from the API.
+   *
+   * @returns {Promise<VerificationDataResponse>} A promise that resolves to the verification data response.
+   * @throws {RequestError} If there is an issue with fetching the data.
+   */
+  async data(): Promise<VerificationDataResponse> {
+    const headers = await this.prepareHeaders();
+    return (await send("GET", `${this.apiDomain}/data`, "", headers)) as VerificationDataResponse;
```
Comment (AhmedHanafy725) [inline packages/grid_client/src/clients/kyc/client.ts]:
```suggestion
    return (await send("GET", `${this.apiDomain}/api/v1/data`, "", headers)) as VerificationDataResponse;
```

Comment (AhmedHanafy725) [reply packages/grid_client/src/clients/kyc/client.ts]: please use url join instead

Comment (AhmedHanafy725) [reply packages/grid_client/src/clients/kyc/client.ts]: please apply this to the rest APIs

File: packages/grid_client/src/clients/kyc/client.ts [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  /**
+   * Prepares the headers required for TFGrid KYC requests.
+   *
+   * This method generates a set of headers that include a timestamp-based challenge,
+   * a signed challenge using the user's key, and other necessary information.
+   *
+   * @returns {Promise<Record<string, string>>} A promise that resolves to an object containing the prepared headers.
+   *
+   */
+  private async prepareHeaders(): Promise<Record<string, string>> {
```
Comment (AhmedHanafy725) [inline packages/grid_client/src/clients/kyc/client.ts:75]: why don't we use `KycHeaders` as the return type?

**Review by AhmedHanafy725 — CHANGES_REQUESTED** [review]

File: packages/grid_client/src/clients/kyc/client.ts (hunk 5 above)
Comment (0oM4R) [reply packages/grid_client/src/clients/kyc/client.ts:75]: the send method accepts `Record<string,string>` I'm using header type here to make it understandable only

File: packages/playground/src/utils/helpers.ts [hunk 6]
```diff
@@ -100,3 +100,25 @@ export function markAsFromAnotherClient(deployment: any): any {
...
   return deployment;
 }
+
+export function handleKYCError(message: string, error: Error): string {
+  if (error instanceof KycErrors.TFGridKycError) {
+    switch (true) {
+      case error instanceof KycErrors.KycInvalidAddressError:
+        return `${message}: Invalid address, please contact support`;
+      case error instanceof KycErrors.KycInvalidChallengeError:
+        return `${message}: Invalid challenge, please contact support`;
```
Comment (amiraabouhadid) [inline packages/playground/src/utils/helpers.ts]: I don't think user would understand what 'invalid challenge' means. Can we rephrase this to perhaps 'Invalid Credentials' or 'Verification Failed'?

File: packages/grid_client/src/client.ts [hunk 7]
```diff
@@ -245,6 +247,8 @@ class GridClient {
       substrate: substrateURL || `wss://tfchain.${base}/ws`,
       graphql: graphqlURL || `https://graphql.${base}/graphql`,
       activation: activationURL || `https://activation.${base}/activation/activate`,
+      //TODO handle kyc url
```
Comment (amiraabouhadid) [inline packages/grid_client/src/client.ts]: remove this comment

File: packages/grid_client/src/clients/kyc/client.ts [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+    const signedChallenge = this.keypair.sign(bytesFromHex(challenge));
+    const signedMsgHex = Buffer.from(signedChallenge).toString("hex");
+
+    const headers: KycHeaders = {
+      "content-type": "application/json",
+      "X-Client-ID": this.address,
+      "X-Challenge": challenge,
+      "X-Signature": signedMsgHex,
+    };
+    return headers as unknown as Record<string, string>;
```
Comment (amiraabouhadid) [inline packages/grid_client/src/clients/kyc/client.ts:88]: unknown?

File: packages/grid_client/src/clients/kyc/client.ts [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+   * @throws {KycErrors.KycInvalidAddressError} If the error message indicates a malformed address.
+   * @throws {KycErrors.KycInvalidChallengeError} If the error message indicates a malformed or bad challenge.
+   * @throws {KycErrors.KycInvalidSignatureError} If the error message indicates a malformed or bad signature.
+   * @returns {undefined} If the error message does not match any known error patterns.
+   */
+  private ThrowHeadersRelatedError(errorMessage: string) {
+    switch (true) {
+      case errorMessage.includes("malformed address"):
+        throw new KycErrors.KycInvalidAddressError(errorMessage);
+      case errorMessage.includes("malformed challenge") || errorMessage.includes("bad challenge"):
```
Comment (amiraabouhadid) [inline packages/grid_client/src/clients/kyc/client.ts]: since malformed is in all of these error messages, then why not just check for .includes("address"), .includes("challenge"), etc?

File: packages/grid_client/src/clients/kyc/client.ts [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+          "Failed to get token due to: Response does not contain authToken field",
+        );
+      return res.result.authToken;
+    } catch (error) {
+      const messagePrefix = "Failed to get auth token from KYC service";
+      const errorMessage = formatErrorMessage(messagePrefix, error);
+      const statusCode = (error as RequestError).statusCode;
+      this.ThrowHeadersRelatedError(errorMessage);
+      switch (true) {
+        case statusCode === 429:
```
Comment (amiraabouhadid) [inline packages/grid_client/src/clients/kyc/client.ts]: put these statuscodes into an enum

File: packages/grid_client/src/high_level/twinDeploymentHandler.ts [hunk 11]
```diff
@@ -34,6 +36,8 @@ class TwinDeploymentHandler {
       }
     }
     try {
+      const kycStatus = await this.kyc.status();
+      if (kycStatus !== KycStatus.verified) throw new ValidationError("KYC is not verified.");
```
Comment (amiraabouhadid) [inline packages/grid_client/src/high_level/twinDeploymentHandler.ts]: might be better as "Verification Failed", user doesn't necessarily know what KYC is.

File: packages/grid_client/src/client.ts (hunk 7 above)
Comment (0oM4R) [reply packages/grid_client/src/client.ts]: Needed to be handled later, on supporting kyc stacks

File: packages/grid_client/src/clients/kyc/client.ts (hunk 8 above)
Comment (0oM4R) [reply packages/grid_client/src/clients/kyc/client.ts:88]: Yes we have to convert the type to record so we have to use it [ref](https://stackoverflow.com/questions/69399211/typescript-why-does-as-unknown-as-x-work)
What do you suggest?

**Review by samaradel — COMMENTED** [review]
[Screencast from 23-10-24 15:34:39.webm](https://github.com/user-attachments/assets/902c34fe-7130-4072-9167-3424000b780c)
it keeps loading

**Comment by 0oM4R** [comment]
> [Screencast from 23-10-24 15:34:39.webm](https://github.com/user-attachments/assets/902c34fe-7130-4072-9167-3424000b780c) it keeps loading

did you build the clients ?

**Comment by samaradel** [comment]
@0oM4R yes I did

**Comment by samaradel** [comment]
> [Screencast from 23-10-24 15:34:39.webm](https://github.com/user-attachments/assets/902c34fe-7130-4072-9167-3424000b780c) it keeps loading

Update: It was from ```react extension``` was enabled and it works when it disabled.

File: packages/grid_client/src/clients/kyc/client.ts (hunk 9 above)
Comment (0oM4R) [reply packages/grid_client/src/clients/kyc/client.ts]: the idea is those errors for handling corrupted headers. service may have another error that contains `address` and `challenge` so we need to make sure about the error message context.

File: packages/playground/src/utils/helpers.ts (hunk 6 above)
Comment (0oM4R) [reply packages/playground/src/utils/helpers.ts]: As I discussed with @AhmedHanafy725, those parts of the message is for us as developers and support, we need to know what happen exactly, so we mentioned to contact support

**Review by samaradel — APPROVED** [review]

File: packages/grid_client/src/clients/kyc/client.ts [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+
+      case status === HttpStatusCode.TooManyRequests:
+        return new KycErrors.RateLimit(errorMessage);
+
+      case status === HttpStatusCode.Conflict:
+        return new KycErrors.AlreadyVerified(errorMessage);
+
+      case status === HttpStatusCode.PaymentRequired:
+        return new InsufficientBalanceError(errorMessage);
+    }
```
Comment (xmonader) [inline packages/grid_client/src/clients/kyc/client.ts:123]: And else?

File: packages/types/src/errors/grid_client/kyc.ts [hunk 13]
```diff
@@ -0,0 +1,53 @@
+import { BaseError } from "../base_error";
+import { ErrorModules } from "../modules";
+
+export enum Errors {
+  Unverified,
+  InvalidResponse,
+  RateLimit,
+  badRequest,
```
Comment (xmonader) [inline packages/types/src/errors/grid_client/kyc.ts]: Why the inconsistency? Change to BadRequest, Unauthorized!

File: packages/grid_client/src/high_level/twinDeploymentHandler.ts [hunk 14]
```diff
@@ -502,6 +509,11 @@ class TwinDeploymentHandler {
   }
 
   async handle(twinDeployments: TwinDeployment[]) {
+    const kycStatus = await this.kyc.status();
+    if (kycStatus !== KycStatus.verified)
+      throw new ValidationError(
+        "Your account is not verified. Please sign into Threefold Dashboard or Connect mobile app to complete your KYC verification.",
```
Comment (xmonader) [inline packages/grid_client/src/high_level/twinDeploymentHandler.ts:510]: Revise the app name it's called `Threefold Connect`

Comment (0oM4R) [reply packages/grid_client/src/high_level/twinDeploymentHandler.ts:510]: replaced with ` Your account is not verified. Please sign into ThreeFold Dashboard or ThreeFold Connect mobile app to complete your KYC verification.`

File: packages/types/src/errors/grid_client/kyc.ts (hunk 13 above)
Comment (0oM4R) [reply packages/types/src/errors/grid_client/kyc.ts]: updated; thank you!

File: packages/grid_client/src/clients/kyc/client.ts (hunk 12 above)
Comment (0oM4R) [reply packages/grid_client/src/clients/kyc/client.ts:123]: default case was removed by mistake, added now

File: packages/grid_client/src/high_level/twinDeploymentHandler.ts [hunk 15]
```diff
@@ -34,6 +36,11 @@ class TwinDeploymentHandler {
       }
     }
     try {
+      const kycStatus = await this.kyc.status();
+      if (kycStatus !== KycStatus.verified)
+        throw new ValidationError(
+          "Your account is not verified. Please sign into ThreeFold Dashboard or ThreeFold Connect mobile app to complete your KYC verification.",
+        ); //TODO add kyc manual link
```
Comment (AhmedHanafy725) [inline packages/grid_client/src/high_level/twinDeploymentHandler.ts]: this is not required, already domains require 2 contracts (name contract to be unique and node contract to deploy it)

**Review by AhmedHanafy725 — APPROVED** [review]

**Review by xmonader — APPROVED** [review]
